fix: opt in checkbox to respect the user choice when previously opted out

rjchow opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR took on a life of its own and turned out to be a fair bit more complex to implement than originally intended, as it seems that `woocommerce_allow_tracking` is set to 'no' by default when WooCommerce is installed.

This made it necessary to implement a way to determine if the user has completed the Opt-in step (and/or the entire Core Profiler) previously. This was implemented by storing an object within `woocommerce_onboarding_profile_progress` under the key: `core_profiler_completed_steps`. 

The reason it has a second level of `core_profiler_completed_steps` instead of directly storing the steps is because it leaves open the possibility of storing different step completions for different onboarding flows, e.g for mobile

This object looks like: 

```
{
	"core_profiler_completed_steps": {
		"intro-opt-in": {
			"completed_at": "2024-11-12T06:50:21Z"
		},
		"user-profile": {
			"completed_at": "2024-11-12T06:50:35Z"
		},
		"business-info": {
			"completed_at": "2024-11-12T06:47:24Z"
		}
	}
}
```

Part of implementing this involved adding new endpoints for setting steps to complete, and while implementing this I realised that there was a race condition caused by two API calls both writing to the `woocommerce_onboarding_profile` option. Thus I've implemented ~a mutex to ensure that there's no data loss during a race condition.~ it in a different option instead of within `woocommerce_onboarding_profile`. 

(I could not get the mutex to work reliably due to WP object caching on the host side)



Closes https://github.com/woocommerce/woocommerce/issues/49895.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

There should be 3 scenarios to test:
1. Site is new and hence user has not completed the Core Profiler or Opted In => Checkbox should be checked by default
2. User has previously opted out, so `woocommerce_allow_tracking` should be 'no' => Checkbox should not be checked
3. User has previously opted in, so `woocommerce_allow_tracking` should be 'yes' => Checkbox should be checked

To test this, 
1. Set up WooCommerce along with WC Beta Tester plugin for easy access to change the `woocommerce_allow_tracking` option
2. Go through the Core Profiler onboarding, and on the first page there is a checkbox indicating usage tracking opt-in.
3. Assert the above behaviours hold. You can delete the options (`woocommerce_allow_tracking` and `woocommerce_onboarding_profile` using the WC admin test helper.
4. These behaviours should hold even through a refresh or using the forward/back browser buttons.

<img width="983" alt="image" src="https://github.com/user-attachments/assets/a60d099c-a177-430f-ba41-76f7ca1fdd10">



<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

rjchow commented:
Converting to draft as it doesn't work as intended because the 'woocommerce_allow_tracking' option is initialised to 'no' as part of setup

rjchow commented on plugins/woocommerce/src/Admin/API/OnboardingProfile.php:
```diff
@@ -153,11 +224,26 @@ protected function filter_industries( $industries ) {
+
+		self::release_lock();
+
+		/**
```
Documentation update triggered by lint running on modified file

rjchow commented on plugins/woocommerce/src/Admin/API/OnboardingProfile.php:
```diff
@@ -136,10 +201,16 @@ public function get_items( $request ) {
 	 * @return array
 	 */
 	protected function filter_industries( $industries ) {
+		/**
```
Documentation update triggered by lint running on modified file

psealock commented:
Not sure if its related, but I'm seeing this error a few times on this branch. I've run out of time today to debug, but will continue in the morning.

<img width="1688" alt="image" src="https://github.com/user-attachments/assets/331d7241-268e-4ea5-8984-1e1b2c60c772">

rjchow commented:
Yep @psealock seems like there's an issue and doubly so the mutex didn't prevent the problem. I'm taking this back to draft, sorry!

rjchow committed:
remove keys for null|undefined in updateProfileItem api call

fix: opt in checkbox to respect the user choice when previously opted out

fix type error

wip

moved completion to events instead of exit

- because otherwise even navigation triggers it

lint

lint

lint2

test

introduced mutex lock

changed timeout to 5s

removed mutex

changelog

re-add mutex

wip

updated to new option

rjchow committed:
types n lint

adrianduffell commented on plugins/woocommerce/client/admin/client/core-profiler/pages/IntroOptIn.tsx:
```diff
@@ -13,16 +13,41 @@ import { Link } from '@woocommerce/components';
+		optInCheckboxInitialStatus
+	);
+
+	// we want the checkbox to be checked iff
```
Typo: iff

rjchow replied:
[iff](https://en.wikipedia.org/wiki/If_and_only_if) 😅 The usage made sense when I started with a smaller list of individual conditions but not anymore

adrianduffell commented on plugins/woocommerce/client/admin/client/core-profiler/pages/IntroOptIn.tsx:
```diff
@@ -13,16 +13,41 @@ import { Link } from '@woocommerce/components';
+	// 1. the user has previously completed the profiler and opted out
+	//  1a. the user has completed the intro-opt-in step previously and opted out
+	// a user has completed the profiler if context.userProfile.completed is true or
+	// context.userProfile.skipped is true
```
❤️

adrianduffell reviewed (commented):
It tested well, nice work @rjchow !

psealock commented:
This doesn't seem to be working correctly when I skip the profiler. Am I testing this correctly?

1. Uncheck opt-in
2. Skip profiler, complete location dropdown
3. Delete the two options
4. Navigate to `/wp-admin/admin.php?page=wc-admin&path=%2Fsetup-wizard`
5. See the opt-in checkbox ticked

rjchow commented:
> This doesn't seem to be working correctly when I skip the profiler. Am I testing this correctly?
> 
>     1. Uncheck opt-in
> 
>     2. Skip profiler, complete location dropdown
> 
>     3. Delete the two options
> 
>     4. Navigate to `/wp-admin/admin.php?page=wc-admin&path=%2Fsetup-wizard`
> 
>     5. See the opt-in checkbox ticked

Hey @psealock thanks for testing! That is the expected behaviour if you delete the options, as deleting them resets the core profiler progress to 'never been seen' which would be scenario 1

For scenario 2 you can revisit the core profiler via URL after skipping or completing it, with the checkbox unticked the first run.

For scenario 3, ditto but with the checkbox ticked the first run.

You would delete the options only between the scenarios

psealock reviewed (approved):
>You would delete the options only between the scenarios

Gotcha, ok this is testing well. And code looks legit too 👍🏽

rjchow committed:
iff -> if
